src/app/chart/donut/donut.component.ts

```diff
@@ -0,0 +1,144 @@
+   * @param chartDefaults
+   */
+  constructor(private chartDefaults: ChartDefaults) {
+    super();
+  }
+
+  ngOnInit(): void {
+    this.donutChartId = 'donutChart';
+    if (this.config.chartId) {
+      this.donutChartId = this.config.chartId + this.donutChartId;
+    }
+
+    this.chartLoaded.subscribe({
+      next: (chart: any) => {
+        this.chartAvailable(chart);
+      }
+    });
+
+    this.setupConfig();
+    super.generateChart(this.donutChartId);
+    this.updateAll();
+  }
+
+  ngDoCheck(): void {
+    if (!isEqual(this.chartData, this.prevChartData)) {
+      this.updateAll();
+    }
+
+    if (!isEqual(this.config, this.prevConfig)) {
+      this.setupConfig();
+      this.generateChart(this.donutChartId, true);
+    }
+  }
+
+  public chartAvailable(chart: any): void {
+    this.setupDonutChartTitle(chart);
+  }
+
+  private getTotal(): number {
+    let total = 0;
+    this.chartData.forEach((element: any) => {
+      if (!isNaN(element[1])) {
+        total += Number(element[1]);
+      }
+    });
+
+    return total;
+  }
+
+  private getCenterLabelText(): any {
+    let centerLabelText = {
+      bigText: this.getTotal(),
+      smText: this.config.donut.title
+    };
+
+    if (this.config.centerLabelFn) {
+      centerLabelText.bigText = this.config.centerLabelFn();
+      centerLabelText.smText = '';
+    }
+
+    return centerLabelText;
+  }
+
+  private setupDonutChartTitle(chart: any): void {
+    let donutChartTitle, centerLabelText;
+
+    if (!chart) {
+      return;
+    }
+
+    donutChartTitle = d3.select(chart.element).select('text.c3-chart-arcs-title');
+    if (!donutChartTitle) {
+      return;
+    }
+
+    centerLabelText = this.getCenterLabelText();
+
+    donutChartTitle.text('');
+    if (centerLabelText.bigText && !centerLabelText.smText) {
+      donutChartTitle.text(centerLabelText.bigText);
+    } else {
+      donutChartTitle.insert('tspan', null).text(centerLabelText.bigText)
+        .classed('donut-title-big-pf', true).attr('dy', 0).attr('x', 0);
+      donutChartTitle.insert('tspan', null).text(centerLabelText.smText).
+        classed('donut-title-small-pf', true).attr('dy', 20).attr('x', 0);
+    }
+  }
+
+  private setupConfig(): void {
+    this.defaultConfig = this.chartDefaults.getDefaultDonutConfig();
+    let defaultDonut = this.chartDefaults.getDefaultDonut();
+
+    defaults(this.config, this.defaultConfig);
+    defaults(this.config.donut, defaultDonut);
+
+    if (this.config.chartHeight) {
+      this.defaultConfig.size.height = this.config.chartHeight;
+      this.config.size.height = this.config.chartHeight;
+    }
+  }
+
+  private getDonutData(chartData: any): any {
+    return {
+      type: 'donut',
+      columns: this.chartData,
+      order: null,
+      colors: this.config.colors
+    };
+  }
+
+  private updateAll(): void {
+    this.prevChartData = cloneDeep(this.chartData);
+
+    this.config.data = merge(this.config.data, this.getDonutData(this.chartData));
```

Comment:
If you're merging data, the config should be cloned after this change. That is, ensure setupConfig is always called after updateAll